### PR TITLE
Off-grid sensors. Band-limited interpolation and tests.

### PR DESCRIPTION
**Pull request by tomelse**
As discussed in the issues: here's the BLI implementation for off-grid sensors. 

I've replicated some of the tests from kWave, which seem to be sanity checks for the most part. I've also added some example code to the documentation.

Let me know if there's anything to add!

### CODE REVIEW

**Comment by astanziola** [comment]
This looks fantastic, thank you very much for putting this PR together!! 
I'll try to do a full review by tomorrow the latest, just dealing with a few deadlines today

I really love the notebook example that solves for the radius of the transducers ring! Really a nice way to show the full differentiability :smile:

File: tests/test_off_grid_sensors.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+from jwave.geometry import _bli_function, BLISensors, Domain, FourierSeries
+import numpy as np
+import pytest
+
+
+@pytest.mark.parametrize("n_grid", [(100, ), (101, )])
+def test_bli_function(n_grid=100):
```
Comment (astanziola) [inline tests/test_off_grid_sensors.py:7]: The current tests fail here: https://github.com/ucl-bug/jwave/actions/runs/5381874402/jobs/9785603397?pr=192#step:5:37

I think it is just because one can't put a default value for the arguments of a parametrized test, so probably changing to `def test_bli_function(ngrid):` will be enough

File: pyproject.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "jwave"
-version = "0.1.2"
+version = "0.1.3"
```
Comment (astanziola) [inline pyproject.toml:3]: This can be kept at `0.1.2`, the github action takes care automatically to upgrade the version at each release.

File: jwave/geometry.py [hunk 3]
```diff
@@ -15,10 +15,12 @@
 
 import math
 from dataclasses import dataclass
+from functools import partial
 from typing import List, Tuple, Union
+from numpy.typing import ArrayLike
 
 import numpy as np
-from jax import numpy as jnp
+from jax import numpy as jnp, custom_jvp
```
Comment (astanziola) [inline jwave/geometry.py:23]: I believe that the `custom_jvp` is not used anymore?

If you want, you can use the `pre-commit` of this repo to automatically take care of unused imports and other similarly annoying things :)
If you have never used it before (and you don't have to!), you can find here how to use it: https://github.com/ucl-bug/jwave/blob/main/CONTRIBUTING.md#getting-started

File: jwave/geometry.py [hunk 4]
```diff
@@ -461,10 +463,102 @@ def __call__(self, p: Field, u: Field, rho: Field):
...
+                "Sensors positions must be 1, 2 or 3 dimensional. Not {}".
+                format(len(self.positions)))
+
+
+def _bli_function(x0, x, n, include_imag=False):
+    # x0 = position of sensors along axis
+    # x = grid
+    # n = size of grid
+    # include_imag = include the imaginary component?
+    dx = jnp.where((x - x0[:, None]) == 0, x - x0[:, None] + 1e-4, x - x0[:, None])  # Stop us getting nans.
```
Comment (astanziola) [inline jwave/geometry.py:481]: Can you help me understand what this `1e-4` term is here?

If that's to avoid division by zero when the point is on grid, we could perhaps do something similar to what is done to implement the `sinc` function in jax: https://github.com/google/jax/blob/6bc74d2a9874e1fe93a45191bb829c07dfee04fa/jax/_src/numpy/ufuncs.py#L717

The idea is to use a series representation when the input is zero (since the only thing we care of is to get the correct gradients).

If you don't have time to do this now no worries at all, just let me know and I will open an issue and tackle it afterwards.

File: jwave/geometry.py [hunk 5]
```diff
@@ -461,10 +463,102 @@ def __call__(self, p: Field, u: Field, rho: Field):
...
+class BLISensors:
+    """ Band-limited interpolant (off-grid) sensors.
+
+    Args:
+        positions (Tuple of List of float): The positions of the sensors.
+        n (Tuple of int): The shape of the grid.
+    """
+
+    positions: Tuple[ArrayLike]
+
```
Comment (astanziola) [inline jwave/geometry.py:509]: Since it seems that you are setting `self.N` in the `__init__` function, it may be worth defining the type of `N` here, in the same way as it is done for `positions`

**Review by astanziola — CHANGES_REQUESTED** [review]
@tomelse  thanks again very much for this PR!
I have marked a few minor things that I believe would be nice to change, but again none of those is urgent and they can all be dealt with in a separate issue / PR. 

Similarly, if you have time you could add an entry to the changelog, something like the following command

```bash
kacl-cli add added "Added off grid sensors [@tomelse] " --modify
```

This is just to make sure that in the next release this will show up in the Changelog. Once again, no worries if you don't have time to do this, I can take care of it :)

Thanks again!

File: jwave/geometry.py (hunk 4 above)
Comment (tomelse) [reply jwave/geometry.py]: I was trying to do something like this: https://github.com/google/jax/issues/1052. I've changed line 481 to the following for clarity:

```python 
    dx = jnp.where((x - x0[:, None]) == 0, 1, x - x0[:, None])  # https://github.com/google/jax/issues/1052
```

**Comment by tomelse** [comment]
Tests are all passing now! https://github.com/tomelse/jwave/actions/runs/5392529223. Let me know if there's anything else you'd like to see changed

**Review by astanziola — APPROVED** [review]
LGTM! Thanks a lot again for this @tomelse :)

File: jwave/geometry.py (hunk 4 above)
Comment (astanziola) [reply jwave/geometry.py]: That's a very interesting sharp bit of jax that I didn't have the pleasure of seeing yet! 😄
